plugin/hosts/setup.go

```diff
@@ -22,28 +23,37 @@ func init() {
 	})
 }
 
+func periodicHostsUpdate(h *Hosts) chan bool {
+	parseChan := make(chan bool)
+
+	if h.options.reload == durationOf0s {
+		return parseChan
+	}
+
+	go func() {
+		ticker := time.NewTicker(h.options.reload)
+		for {
+			select {
+			case <-parseChan:
+				return
+			case <-ticker.C:
+				h.readHosts()
+			}
+		}
+	}()
+	return parseChan
+}
+
 func setup(c *caddy.Controller) error {
 	h, err := hostsParse(c)
 	if err != nil {
 		return plugin.Error("hosts", err)
 	}
 
-	parseChan := make(chan bool)
+	parseChan := periodicHostsUpdate(&h)
 
 	c.OnStartup(func() error {
 		h.readHosts()
-
-		go func() {
```

Comment:
why can't we just keep this structure and just not execute it when the duration is 0?

Reply:
If you are commenting about the un-inlining, I could re-inline the code but I thought having a function naming the feature improved readability and simplified the code flow of the main `setup` function.

I thought about the code structure and decided for this one as to not require us to perform the test both in the `OnStartup` and `OnShutdown`. This way the test is within the code block but from the outside the behaviour is consistent.

What do you had in mind as change ?

Reply:
fair enough